docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+unnecessary, but it will reduce SnippetsPolicy applicability.
+
+Although a SnippetsPolicy can target different Kinds (like Gateway and HTTPRoute), considering that
+(1) each SnippetPolicy is independent of any other and (2) a single SnippetPolicy can only affect resources of the
+same Kind, SnippetsPolicy is a [Direct Attached Policy](https://gateway-api.sigs.k8s.io/geps/gep-2648/).
+
+### Policy or Filter
+
+Considering that SnippetPolicy can be used to implement authentication and authorization, because of the reasons
+mentioned
+[here](nginx-extensions.md#authentication) it makes sense to also introduce SnippetsFilter, which will be the same as
+SnippetsPolicy but without the `targetRefs` field:
+
+```yaml
+apiVersion: gateway.nginx.org/v1alpha1
+kind: SnippetsFilter
+metadata:
+  name: access-control
+spec:
+  snippets:
+  - context: http.server
+    value: |
+      allow 10.0.0.0/8;
+      deny all;
+```
+
+An HTTPRoute can include that filter:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: HTTPRoute
+metadata:
+  name: my-app
+spec:
+  . . .
+  rules:
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /
+    filters:
+    - type: ExtensionRef
+      extensionRef:
+        group: gateway.nginx.org/v1alpha1
+        kind: SnippetsFilter
+        name: access-control
+    backendRefs:
+    - name: headers
+      port: 80
+```
+
+SnippetsFilter can also be used in
+GRPCRoute. [TLSRoute](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1alpha2.TLSRoute)
+doesn't support filters, so it needs to be added to the Gateway API first.
+
+Although SnippetsFilter applicability is limited compared to SnippetsPolicy, SnippetsFilter creates
+a natural split of responsibilities between Cluster operator and Application developers:
+
+- Cluster operator creates a SnippetsFilter.
+- Application developers reference the SnippetsFilter to enable it.
+
+Note that with the SnippetsPolicy, because the targetRef is part of the SnippetsPolicy, it is not possible to have
+such a split of responsibilities.
+
+TO-DO(pleshakov) - Add SnippetsFilter to the GEP after discussions about it.
+
+### Personas
+
+The target persona of snippets is Cluster operator. They will create and manage snippets.
+
+Snippets are not intended for Application developers, because:
+
+- To create snippets, you need to know about NGINX and how NGF generates NGINX configuration.
+- Snippets can easily break NGINX config (if not validated by NGF). See
+  [Config Validation section](#nginx-config-validation) below.
+- Snippets can be used to exploit NGF. (See [Security Considerations section](#security-considerations) below)
+
+As mentioned in the [Policy or Filter](#policy-or-filter), when snippets are used via SnippetsFilter, Application
+developers can still control whether they want to enable or disable snippets by referencing them in an HTTPRoute.
+
+### NGINX Config Validation
```

Comment:
What about validating the policy spec itself -- for example, verifying that the context is allowed by the targetRef?